xds/internal/balancer/gracefulswitch/gracefulswitch_test.go

```diff
@@ -0,0 +1,1354 @@
+		}
+	}
+
+	deadline := time.Now().Add(defaultTestTimeout)
+	// Poll to see if pending was deleted (happens in forked goroutine).
+	for {
+		gsb.mu.Lock()
+		if gsb.balancerPending == nil {
+			gsb.mu.Unlock()
+			break
+		}
+		gsb.mu.Unlock()
+		if time.Now().After(deadline) {
+			t.Fatalf("balancerPending was not deleted as the pending LB reported a state other than READY, which should switch pending to current")
+		}
+		time.Sleep(time.Millisecond)
+	}
+}
+
+// TestCurrentNotReadyPendingUpdate tests the scenario where there is a Current
+// and Pending Load Balancer setup in the Graceful Switch Load Balancer, and the
+// current LB is not in the connectivity state READY. Any update from the
+// pending Load Balancer should cause the Graceful Switch Load Balancer to swap
+// the Pending into Current, and update the ClientConn with the Pending Load
+// Balancers state.
+func (s) TestCurrentNotReadyPendingUpdate(t *testing.T) {
+	tcc, gsb := setup(t)
+	builder := balancer.Get(balancerName1)
+	if builder == nil {
+		t.Fatalf("balancer.Get(%v) returned nil", balancerName1)
+	}
+	gsb.SwitchTo(builder)
+	builder = balancer.Get(balancerName1)
+	if builder == nil {
+		t.Fatalf("balancer.Get(%v) returned nil", balancerName1)
+	}
+	gsb.SwitchTo(builder)
+	if gsb.balancerPending == nil {
+		t.Fatal("balancerPending was not built out when another SwitchTo() call should've triggered the pending balancer to build")
+	}
+	// Due to the Current Load Balancer not being in state READY, any update
+	// from the Pending Load Balancer should cause that update to be forwarded
+	// to the Client Conn and also cause the Pending Load Balancer to swap into
+	// the current one.
+	gsb.balancerPending.Balancer.(*mockBalancer1).updateState(balancer.State{
+		ConnectivityState: connectivity.Connecting,
+		Picker:            &neverErrPicker{},
+	})
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+	defer cancel()
+	select {
+	case <-ctx.Done():
+		t.Fatalf("timeout waiting for an UpdateState call on the Client Conn")
+	case state := <-tcc.NewStateCh:
+		if state != connectivity.Connecting {
+			t.Fatalf("Graceful Switch Balancer should have sent the connectivity state CONNECTING received from recent update")
+		}
+	}
+	ctx, cancel = context.WithTimeout(context.Background(), defaultTestTimeout)
+	defer cancel()
+	select {
+	case <-ctx.Done():
+		t.Fatalf("timeout waiting for an UpdateState call on the Client Conn")
+	case picker := <-tcc.NewPickerCh:
+		// Should receive a never err picker.
+		_, err := picker.Pick(balancer.PickInfo{})
+		if err != nil {
+			t.Fatalf("Graceful Switch Balancer should have sent a never error picker, which is a picker that never errors, instead received err %v", err)
+		}
+	}
+
+	deadline := time.Now().Add(defaultTestTimeout)
+	// Poll to see if pending was deleted (happens in forked goroutine).
```

Comment:
Should these polling loops wait for `Close` to be called on the pending policy instead?  Then they can be blocked on a channel read instead of needing to poll?

Reply:
Switched all polling loops to waiting for Close to be called on the pending policy instead.